bundles/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/internal/handler/PHCBridgeHandler.java

```diff
@@ -104,215 +110,405 @@ public void initialize() {
-            }
+            byte module = -1;
+            while (buffer.size() >= 5) {
+                loopCounter = 0;
 
-        } catch (IOException e) {
-            // ignore
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-        }
-    }
+                if (module == -1) {
+                    module = buffer.get();
+                }
 
-    /**
-     * Split and process the given buffer to single messages.
-     *
-     * @param buffer
-     */
-    private void processInputStream(byte[] buffer) {
-        List<Byte> result = new ArrayList<Byte>();
-        List<String> repeat = new ArrayList<String>();
-        int pos = 0;
-        messageFragment = null;
-        Byte moduleAddress = null;
+                // not a known module address
+                if (!modules.contains(module)) {
+                    module = -1;
+                    continue;
+                }
 
-        while (pos < buffer.length) {
-            int messageStart = 0;
+                if (logger.isDebugEnabled()) {
+                    logger.debug("get module: {}", PHCHelper.byteToHexString(module));
+                }
 
-            for (byte b : buffer) {
-                if (modules.contains(b)) {
-                    moduleAddress = b;
-                    messageStart = pos++;
-                    break;
+                byte sizeToggle = buffer.get();
+
+                if (logger.isDebugEnabled()) {
+                    logger.debug("get sizeToggle: {}", PHCHelper.byteToHexString(sizeToggle));
                 }
-                pos++;
-            }
 
-            if (moduleAddress == null) {
-                return;
-            }
+                // read length of command and check if makes sense
+                if ((sizeToggle < 1 || sizeToggle > 9) && ((sizeToggle & 0xFF) < 0x81 || (sizeToggle & 0xFF) > 0x89)) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("sizeToggle incorrect");
+                    }
 
-            if (buffer.length - 1 - pos < 3) {
-                messageFragment = new byte[buffer.length - pos + 1];
-                System.arraycopy(buffer, messageStart, messageFragment, 0, messageFragment.length);
-            }
+                    module = sizeToggle;
+                    continue;
+                }
+
+                // read toggle, size and command
+                int size = (sizeToggle & 0x7F);
+                boolean toggle = (sizeToggle & 0x80) == 0x80;
+
+                byte[] command = new byte[size];
+
+                try {
+                    for (int i = 0; i < size; i++) {
+                        command[i] = buffer.get();
+                    }
+
+                    // log command
+                    if (logger.isDebugEnabled()) {
+                        String bin = "";
+                        for (byte b : command) {
+                            bin += PHCHelper.byteToBinaryString(b);
```

Comment:
See other comment

Reply:
What do you mean here? Here I do not create a hex but a binary string.

Reply:
Sorry :-)